flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+
+		final TaskExecutorConnection taskManagerConnection = new TaskExecutorConnection(resourceID, taskExecutorGateway);
+
+		final SlotID slotId = new SlotID(resourceID, 0);
+		final SlotReport slotReport = new SlotReport(createFreeSlotStatus(slotId));
+
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec(resourceManagerId, resourceManagerActions)) {
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport);
+
+			slotManager.processResourceRequirements(requirements);
+
+			DeclarativeTaskManagerSlot slot = slotTracker.getSlot(slotId);
+
+			assertThat(slot.getState(), is(SlotState.ALLOCATED));
+
+			slotManager.processResourceRequirements(requirements);
+		}
+
+		// check that we have only called the resource allocation only for the first slot request,
+		// since the second request is a duplicate
+		assertThat(allocateResourceCalls.get(), is(0));
+	}
+
+	/**
+	 * Tests that a slot allocated for one job can be allocated for another job after being freed.
+	 */
+	@Test
+	public void testSlotCanBeAllocatedForDifferentJobAfterFree() throws Exception {
+		final AllocationID allocationId = new AllocationID();
+		final ResourceRequirements resourceRequirements1 = createResourceRequirementsForSingleSlot();
+		final ResourceRequirements resourceRequirements2 = createResourceRequirementsForSingleSlot();
+
+		final TaskExecutorConnection taskManagerConnection = createTaskExecutorConnection();
+		final ResourceID resourceID = taskManagerConnection.getResourceID();
+
+		final SlotID slotId = new SlotID(resourceID, 0);
+		final SlotReport slotReport = new SlotReport(createFreeSlotStatus(slotId));
+
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec()) {
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport);
+
+			slotManager.processResourceRequirements(resourceRequirements1);
+
+			DeclarativeTaskManagerSlot slot = slotTracker.getSlot(slotId);
+
+			assertEquals("The slot has not been allocated to the expected job id.", resourceRequirements1.getJobId(), slot.getJobId());
+
+			// clear resource requirements, so that the slot isn't immediately re-assigned
+			slotManager.processResourceRequirements(ResourceRequirements.create(resourceRequirements1.getJobId(), resourceRequirements1.getTargetAddress(), Collections.emptyList()));
+			slotManager.freeSlot(slotId, allocationId);
+
+			// check that the slot has been freed
+			assertSame(SlotState.FREE, slot.getState());
+
+			slotManager.processResourceRequirements(resourceRequirements2);
+
+			assertEquals("The slot has not been allocated to the expected job id.", resourceRequirements2.getJobId(), slot.getJobId());
+		}
+	}
```

Comment:
Should we also test that the slo is being used to fulfill requirements of another job if the resource requirements of this job were already registered?

Reply:
that's a good idea, will add it